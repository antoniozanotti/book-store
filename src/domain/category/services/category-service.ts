import { CategoryWithPageAndProductsPrisma, CategoryWithPagePrisma } from "../types/category-prisma";
import prisma from "@/lib/db";

export async function getCategoryWithPageAndProductsBySlug(
  slug: string
): Promise<CategoryWithPageAndProductsPrisma> {
  return await prisma.category.findFirstOrThrow({
    where: {
      page: {
        slug: slug,
      },
    },
    include: {
      page: true,
      products: {
        include: {
          book: {
            include: {
              authors: true,
            },
          },
          page: true,
        },
      },
    },
  });
}

export async function getAllCategoriesWithPage(): Promise<CategoryWithPagePrisma[]> {
  return await prisma.category.findMany({
    include: {
      page: true,
    },
  });
}
