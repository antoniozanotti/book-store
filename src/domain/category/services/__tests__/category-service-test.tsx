/**
 * @jest-environment node
 */
import { categoriesWithPageMock } from "../../__tests__/categories-with-page-mock";
import { getAllCategoriesWithPage, getCategoryWithPageAndProductsBySlug } from "../category-service";

describe("CategoryService", () => {
  it("should return categories with page", async () => {
    const categories = await getAllCategoriesWithPage();
    expect(categories).toMatchSnapshot();
  });

  it("should return category with page and products by slug", async () => {
    const categories = await getCategoryWithPageAndProductsBySlug(categoriesWithPageMock[0].page.slug);
    expect(categories).toMatchSnapshot();
  });
});