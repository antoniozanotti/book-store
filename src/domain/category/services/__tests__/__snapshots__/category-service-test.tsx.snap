// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryService should return categories with page 1`] = `
[
  {
    "description": "Stipes repellat cupiditate vestrum ago consequatur sunt acsi angustus aegre. Aequus eveniet decor vesica succedo laboriosam fugit. Aro aeternus suffoco cometes.",
    "id": 1,
    "name": "Arts and Entertainment",
    "page": {
      "description": "Stipes repellat cupiditate vestrum ago consequatur sunt acsi angustus aegre. Aequus eveniet decor vesica succedo laboriosam fugit. Aro aeternus suffoco comet...",
      "id": 15,
      "slug": "arts-and-entertainment",
      "title": "Arts and Entertainment Books",
      "type": "category",
    },
    "pageId": 15,
  },
  {
    "description": "Amplexus acervus balbus porro quo avarus administratio uxor umbra. Vita decretum urbs sequi. Temeritas deludo turbo demulceo abutor.",
    "id": 2,
    "name": "Business and Technology",
    "page": {
      "description": "Amplexus acervus balbus porro quo avarus administratio uxor umbra. Vita decretum urbs sequi. Temeritas deludo turbo demulceo abutor.",
      "id": 48,
      "slug": "business-and-technology",
      "title": "Business and Technology Books",
      "type": "category",
    },
    "pageId": 48,
  },
  {
    "description": "Verumtamen pecto paulatim provident demitto pariatur appello defero. Aeger usitas denique admoneo charisma abduco adflicto soluta atrocitas vapulus. Thesis cena communis ut molestiae advenio sordeo molestias.",
    "id": 3,
    "name": "Graphic Novels",
    "page": {
      "description": "Verumtamen pecto paulatim provident demitto pariatur appello defero. Aeger usitas denique admoneo charisma abduco adflicto soluta atrocitas vapulus. Thesis c...",
      "id": 80,
      "slug": "graphic-novels",
      "title": "Graphic Novels Books",
      "type": "category",
    },
    "pageId": 80,
  },
  {
    "description": "Iusto illo spes advenio sustineo subnecto degero. Provident amo officiis bellicus cornu esse qui certus. Arceo thesaurus cinis tres debitis color tabernus quod deleniti.",
    "id": 4,
    "name": "History, Biography, and Social Science",
    "page": {
      "description": "Iusto illo spes advenio sustineo subnecto degero. Provident amo officiis bellicus cornu esse qui certus. Arceo thesaurus cinis tres debitis color tabernus qu...",
      "id": 115,
      "slug": "history-biography-and-social-science",
      "title": "History, Biography, and Social Science Books",
      "type": "category",
    },
    "pageId": 115,
  },
  {
    "description": "Spes possimus viridis ratione alveus totidem. Decerno correptius ulciscor amicitia cunctatio coniecto tendo volaticus. Tabesco sopor absque numquam taceo cernuus tersus adstringo minima.",
    "id": 5,
    "name": "Home and Lifestyle",
    "page": {
      "description": "Spes possimus viridis ratione alveus totidem. Decerno correptius ulciscor amicitia cunctatio coniecto tendo volaticus. Tabesco sopor absque numquam taceo cer...",
      "id": 146,
      "slug": "home-and-lifestyle",
      "title": "Home and Lifestyle Books",
      "type": "category",
    },
    "pageId": 146,
  },
  {
    "description": "Vinitor cedo astrum terminatio. Studio calculus vestigium desparatus cibus. Auxilium adipisci studio.",
    "id": 6,
    "name": "Horror",
    "page": {
      "description": "Vinitor cedo astrum terminatio. Studio calculus vestigium desparatus cibus. Auxilium adipisci studio.",
      "id": 178,
      "slug": "horror",
      "title": "Horror Books",
      "type": "category",
    },
    "pageId": 178,
  },
  {
    "description": "Claro voveo defetiscor voluptatum cernuus. Templum vulgus quidem aliqua. Auxilium aiunt absens bis temperantia armarium sumptus.",
    "id": 7,
    "name": "Languages and Travel",
    "page": {
      "description": "Claro voveo defetiscor voluptatum cernuus. Templum vulgus quidem aliqua. Auxilium aiunt absens bis temperantia armarium sumptus.",
      "id": 209,
      "slug": "languages-and-travel",
      "title": "Languages and Travel Books",
      "type": "category",
    },
    "pageId": 209,
  },
  {
    "description": "Subiungo autem alter ustulo comitatus vulgus conqueror adicio custodia defessus. Cui arx custodia deludo. Venio consequuntur vitiosus quas crastinus artificiose aeternus depulso desparatus aveho.",
    "id": 8,
    "name": "LGBTQIA+",
    "page": {
      "description": "Subiungo autem alter ustulo comitatus vulgus conqueror adicio custodia defessus. Cui arx custodia deludo. Venio consequuntur vitiosus quas crastinus artifici...",
      "id": 240,
      "slug": "lgbtqia",
      "title": "LGBTQIA+ Books",
      "type": "category",
    },
    "pageId": 240,
  },
  {
    "description": "Placeat beatus statim claudeo. Ademptio aspernatur maxime dolorum cohibeo adnuo vesco somnus demoror ascisco. Expedita atrox decimus suppellex atqui at valetudo validus universe sono.",
    "id": 9,
    "name": "Literature and Poetry",
    "page": {
      "description": "Placeat beatus statim claudeo. Ademptio aspernatur maxime dolorum cohibeo adnuo vesco somnus demoror ascisco. Expedita atrox decimus suppellex atqui at valet...",
      "id": 271,
      "slug": "literature-and-poetry",
      "title": "Literature and Poetry Books",
      "type": "category",
    },
    "pageId": 271,
  },
  {
    "description": "Deduco aliquid theologus adstringo. Cognatus pecto ubi defungo arbitro. Thermae tamen necessitatibus adopto coma inventore cruciamentum deprecator celo.",
    "id": 10,
    "name": "Mystery",
    "page": {
      "description": "Deduco aliquid theologus adstringo. Cognatus pecto ubi defungo arbitro. Thermae tamen necessitatibus adopto coma inventore cruciamentum deprecator celo.",
      "id": 302,
      "slug": "mystery",
      "title": "Mystery Books",
      "type": "category",
    },
    "pageId": 302,
  },
  {
    "description": "Modi acquiro coniuratio caelestis textilis suffoco ater. Spiritus eos absens. Tepidus error repellat canonicus cerno.",
    "id": 11,
    "name": "Romance",
    "page": {
      "description": "Modi acquiro coniuratio caelestis textilis suffoco ater. Spiritus eos absens. Tepidus error repellat canonicus cerno.",
      "id": 333,
      "slug": "romance",
      "title": "Romance Books",
      "type": "category",
    },
    "pageId": 333,
  },
  {
    "description": "Caecus sapiente coerceo ventosus. Auctor quia dicta odio annus. Tamdiu nihil sufficio verbera super ulterius nihil audax.",
    "id": 12,
    "name": "Science and Math",
    "page": {
      "description": "Caecus sapiente coerceo ventosus. Auctor quia dicta odio annus. Tamdiu nihil sufficio verbera super ulterius nihil audax.",
      "id": 364,
      "slug": "science-and-math",
      "title": "Science and Math Books",
      "type": "category",
    },
    "pageId": 364,
  },
  {
    "description": "Solitudo vaco tandem. Cubo validus decor supplanto tendo umbra vester comes terga. Confugo cedo thymum aegrus.",
    "id": 13,
    "name": "Science Fiction and Fantasy",
    "page": {
      "description": "Solitudo vaco tandem. Cubo validus decor supplanto tendo umbra vester comes terga. Confugo cedo thymum aegrus.",
      "id": 395,
      "slug": "science-fiction-and-fantasy",
      "title": "Science Fiction and Fantasy Books",
      "type": "category",
    },
    "pageId": 395,
  },
  {
    "description": "Thema tripudio aliquid eum. Enim stultus bis vetus pecus tersus. Debilito nihil demulceo tepesco adhuc admiratio tremo timor.",
    "id": 14,
    "name": "Spirituality and Philosophy",
    "page": {
      "description": "Thema tripudio aliquid eum. Enim stultus bis vetus pecus tersus. Debilito nihil demulceo tepesco adhuc admiratio tremo timor.",
      "id": 426,
      "slug": "spirituality-and-philosophy",
      "title": "Spirituality and Philosophy Books",
      "type": "category",
    },
    "pageId": 426,
  },
  {
    "description": "Crustulum comptus comprehendo. Crinis corpus cuius absum summopere aegrotatio. Occaecati spoliatio adeptio cui supellex cupiditas balbus aperte.",
    "id": 15,
    "name": "Wellness, Sports, and Outdoors",
    "page": {
      "description": "Crustulum comptus comprehendo. Crinis corpus cuius absum summopere aegrotatio. Occaecati spoliatio adeptio cui supellex cupiditas balbus aperte.",
      "id": 457,
      "slug": "wellness-sports-and-outdoors",
      "title": "Wellness, Sports, and Outdoors Books",
      "type": "category",
    },
    "pageId": 457,
  },
  {
    "description": "Decretum alius statua varius paens utrum bestia capio creptio. Creta amitto studio degenero ulterius vergo valde. Vorago coma bis comminor vae considero depulso sol tergiversatio thermae.",
    "id": 16,
    "name": "Young Adult",
    "page": {
      "description": "Decretum alius statua varius paens utrum bestia capio creptio. Creta amitto studio degenero ulterius vergo valde. Vorago coma bis comminor vae considero depu...",
      "id": 488,
      "slug": "young-adult",
      "title": "Young Adult Books",
      "type": "category",
    },
    "pageId": 488,
  },
]
`;

exports[`CategoryService should return category with page and products by slug 1`] = `
{
  "description": "Stipes repellat cupiditate vestrum ago consequatur sunt acsi angustus aegre. Aequus eveniet decor vesica succedo laboriosam fugit. Aro aeternus suffoco cometes.",
  "id": 1,
  "name": "Arts and Entertainment",
  "page": {
    "description": "Stipes repellat cupiditate vestrum ago consequatur sunt acsi angustus aegre. Aequus eveniet decor vesica succedo laboriosam fugit. Aro aeternus suffoco comet...",
    "id": 15,
    "slug": "arts-and-entertainment",
    "title": "Arts and Entertainment Books",
    "type": "category",
  },
  "pageId": 15,
  "products": [
    {
      "book": {
        "authors": [
          {
            "description": "Arbor color cras trepide commodi illo dolores porro harum. Versus error maxime substantia conor animadverto traho spiculum sed. Vulgivagus amo supra solio correptius strues corroboro speculum.",
            "id": 1,
            "name": "Kevin Kwan",
            "pageId": 18,
          },
        ],
        "id": 1,
        "pages": 443,
        "productId": 1,
        "publisherId": 1,
        "year": 2018,
      },
      "description": "Carbo suscipit stips eligendi ago canonicus coepi. Caput suppellex tondeo auctor tabgo asperiores. Argumentum speciosus nulla antea aranea astrum turbo thorax.",
      "id": 1,
      "isbn": "978-1-7900-7991-9",
      "name": "Crazy Rich Asians",
      "page": {
        "description": "Carbo suscipit stips eligendi ago canonicus coepi. Caput suppellex tondeo auctor tabgo asperiores. Argumentum speciosus nulla antea aranea astrum turbo thora...",
        "id": 16,
        "slug": "crazy-rich-asians",
        "title": "Book Crazy Rich Asians",
        "type": "product",
      },
      "pageId": 16,
      "price": "29",
    },
    {
      "book": {
        "authors": [
          {
            "description": "Tero versus aliquid tempus argumentum blanditiis angulus quibusdam baiulus quia. Corporis coniecto amet ab laborum conculco iusto vigilo. Accusamus adhuc voluptate virtus teneo cedo.",
            "id": 2,
            "name": "Julia Cameron",
            "pageId": 21,
          },
        ],
        "id": 2,
        "pages": 184,
        "productId": 2,
        "publisherId": 2,
        "year": 1997,
      },
      "description": "Cupressus verecundia surgo tactus. Solium ait tergeo benevolentia cur constans vae vapulus speculum dignissimos. Audax arceo concedo venustas volaticus corporis aedificium mollitia contabesco.",
      "id": 2,
      "isbn": "978-0-12-180727-6",
      "name": "The Artist's Way: A Spiritual Path to Higher Creativity",
      "page": {
        "description": "Cupressus verecundia surgo tactus. Solium ait tergeo benevolentia cur constans vae vapulus speculum dignissimos. Audax arceo concedo venustas volaticus corpo...",
        "id": 19,
        "slug": "the-artists-way-a-spiritual-path-to-higher-creativity",
        "title": "Book The Artist's Way: A Spiritual Path to Higher Creativ...",
        "type": "product",
      },
      "pageId": 19,
      "price": "78",
    },
    {
      "book": {
        "authors": [
          {
            "description": "Terminatio vero terra tot. Ara provident adopto facere advenio adiuvo bis teres apparatus. Arbustum sulum absens asper cerno arguo.",
            "id": 3,
            "name": "David Byrne",
            "pageId": 24,
          },
        ],
        "id": 3,
        "pages": 460,
        "productId": 3,
        "publisherId": 3,
        "year": 1994,
      },
      "description": "Combibo soluta contra celebrer. Vado convoco cursim vicinus cogo turba undique ventus valde somnus. Paulatim verumtamen culpa collum utique tenax quisquam considero molestias.",
      "id": 3,
      "isbn": "978-1-148-42632-7",
      "name": "How Music Works",
      "page": {
        "description": "Combibo soluta contra celebrer. Vado convoco cursim vicinus cogo turba undique ventus valde somnus. Paulatim verumtamen culpa collum utique tenax quisquam co...",
        "id": 22,
        "slug": "how-music-works",
        "title": "Book How Music Works",
        "type": "product",
      },
      "pageId": 22,
      "price": "120",
    },
    {
      "book": {
        "authors": [
          {
            "description": "Mollitia alius uter. Auditor deleo vacuus amicitia virtus super. Vesica sto derelinquo uxor deserunt succurro caveo una.",
            "id": 4,
            "name": "Donna Tartt",
            "pageId": 27,
          },
        ],
        "id": 4,
        "pages": 258,
        "productId": 4,
        "publisherId": 4,
        "year": 2015,
      },
      "description": "Depulso textor degusto video crapula circumvenio viriliter admoneo venio. Delicate aspernatur vesco cedo nisi abeo similique causa. Supellex copiose suggero villa conventus auctus.",
      "id": 4,
      "isbn": "978-0-555-38623-1",
      "name": "The Goldfinch",
      "page": {
        "description": "Depulso textor degusto video crapula circumvenio viriliter admoneo venio. Delicate aspernatur vesco cedo nisi abeo similique causa. Supellex copiose suggero ...",
        "id": 25,
        "slug": "the-goldfinch",
        "title": "Book The Goldfinch",
        "type": "product",
      },
      "pageId": 25,
      "price": "378",
    },
    {
      "book": {
        "authors": [
          {
            "description": "Verbum harum ex argumentum vehemens alienus degero animus aeternus. Crapula alius verus curatio texo. Suggero ambitus tepesco patria certus.",
            "id": 5,
            "name": "Jeremy McCarter",
            "pageId": 30,
          },
          {
            "description": "Damno sortitus trucido atque auctor bis maiores abundans. Officiis impedit dapifer expedita universe sollicito derideo suus. Arbustum comminor taedium aureus mollitia suadeo spiritus.",
            "id": 6,
            "name": "Lin-Manuel Miranda",
            "pageId": 31,
          },
        ],
        "id": 5,
        "pages": 199,
        "productId": 5,
        "publisherId": 5,
        "year": 1997,
      },
      "description": "Communis canto comedo cogito. Acer callide quam curiositas. Turpis ceno alienus arx derelinquo titulus.",
      "id": 5,
      "isbn": "978-1-125-85293-4",
      "name": "Hamilton: The Revolution",
      "page": {
        "description": "Communis canto comedo cogito. Acer callide quam curiositas. Turpis ceno alienus arx derelinquo titulus.",
        "id": 28,
        "slug": "hamilton-the-revolution",
        "title": "Book Hamilton: The Revolution",
        "type": "product",
      },
      "pageId": 28,
      "price": "224",
    },
    {
      "book": {
        "authors": [
          {
            "description": "Degenero deripio arcesso. Vulnero terminatio patior. Vesco conor vestigium sit.",
            "id": 7,
            "name": "Patti Smith",
            "pageId": 34,
          },
        ],
        "id": 6,
        "pages": 280,
        "productId": 6,
        "publisherId": 6,
        "year": 2006,
      },
      "description": "Adimpleo sursum accommodo natus. Adicio curia carmen canto defessus conturbo synagoga apto. Correptius vulpes defleo deludo.",
      "id": 6,
      "isbn": "978-0-374-92156-9",
      "name": "Just Kids",
      "page": {
        "description": "Adimpleo sursum accommodo natus. Adicio curia carmen canto defessus conturbo synagoga apto. Correptius vulpes defleo deludo.",
        "id": 32,
        "slug": "just-kids",
        "title": "Book Just Kids",
        "type": "product",
      },
      "pageId": 32,
      "price": "439",
    },
    {
      "book": {
        "authors": [
          {
            "description": "Inventore aut ascit. Acervus despecto repellat agnosco praesentium alias eos coepi. Sortitus decor excepturi cuius thymum turpis conservo.",
            "id": 8,
            "name": "Austin Kleon",
            "pageId": 37,
          },
        ],
        "id": 7,
        "pages": 191,
        "productId": 7,
        "publisherId": 7,
        "year": 1996,
      },
      "description": "Tergo venustas vester utique subseco temporibus pecto textus bonus textilis. Paulatim conqueror doloremque. Curo creber adipisci vinculum.",
      "id": 7,
      "isbn": "978-1-4597-0735-1",
      "name": "Steal Like an Artist: 10 Things Nobody Told You About Being Creative",
      "page": {
        "description": "Tergo venustas vester utique subseco temporibus pecto textus bonus textilis. Paulatim conqueror doloremque. Curo creber adipisci vinculum.",
        "id": 35,
        "slug": "steal-like-an-artist-10-things-nobody-told-you-about-being-creative",
        "title": "Book Steal Like an Artist: 10 Things Nobody Told You Abou...",
        "type": "product",
      },
      "pageId": 35,
      "price": "308",
    },
    {
      "book": {
        "authors": [
          {
            "description": "Versus trepide colligo caveo conscendo aestas adduco. Copiose varius utique carpo tabula absque amplitudo. Antepono canis ancilla capitulus eos spiculum desipio.",
            "id": 9,
            "name": "Bruce Springsteen",
            "pageId": 40,
          },
        ],
        "id": 8,
        "pages": 383,
        "productId": 8,
        "publisherId": 8,
        "year": 2010,
      },
      "description": "Ex eligendi sufficio dedico. Tergum contra verto tam claro arguo aedificium veritas audax amitto. Urbanus supra abstergo ait denuncio vomito fugiat.",
      "id": 8,
      "isbn": "978-1-62672-977-3",
      "name": "Born to Run",
      "page": {
        "description": "Ex eligendi sufficio dedico. Tergum contra verto tam claro arguo aedificium veritas audax amitto. Urbanus supra abstergo ait denuncio vomito fugiat.",
        "id": 38,
        "slug": "born-to-run",
        "title": "Book Born to Run",
        "type": "product",
      },
      "pageId": 38,
      "price": "410",
    },
    {
      "book": {
        "authors": [
          {
            "description": "Cuppedia uberrime auctus coma. Clarus turba paens cervus aufero tripudio vestigium. Repellat appono ara commemoro tollo velut agnosco.",
            "id": 10,
            "name": "John Lasseter",
            "pageId": 43,
          },
          {
            "description": "Agnosco somnus demergo aut copia labore constans. Omnis templum comedo contigo tametsi arma sophismata adsuesco ea patrocinor. Vivo minus umquam.",
            "id": 11,
            "name": "Amid Amidi",
            "pageId": 44,
          },
        ],
        "id": 9,
        "pages": 95,
        "productId": 9,
        "publisherId": 9,
        "year": 1997,
      },
      "description": "Tergum caste sequi reiciendis. Causa omnis absorbeo ea allatus beneficium viriliter decor utpote placeat. Corporis capto supellex curriculum repellat adsuesco tamquam torqueo.",
      "id": 9,
      "isbn": "978-1-889216-25-6",
      "name": "The Art of Pixar: 25th Anniversary",
      "page": {
        "description": "Tergum caste sequi reiciendis. Causa omnis absorbeo ea allatus beneficium viriliter decor utpote placeat. Corporis capto supellex curriculum repellat adsuesc...",
        "id": 41,
        "slug": "the-art-of-pixar-25th-anniversary",
        "title": "Book The Art of Pixar: 25th Anniversary",
        "type": "product",
      },
      "pageId": 41,
      "price": "386",
    },
    {
      "book": {
        "authors": [
          {
            "description": "Spargo clibanus claro vinitor coepi antiquus demonstro. Molestias adstringo ceno thorax spiritus antea in numquam appositus temeritas. Cariosus solum vinitor patria agnosco demoror vicissitudo.",
            "id": 12,
            "name": "Elizabeth Currid-Halkett",
            "pageId": 47,
          },
        ],
        "id": 10,
        "pages": 344,
        "productId": 10,
        "publisherId": 10,
        "year": 1995,
      },
      "description": "Coaegresco a tandem suffoco. Dens coaegresco sint trado chirographum accedo. Taedium tibi utique.",
      "id": 10,
      "isbn": "978-0-06-777614-8",
      "name": "The Warhol Economy",
      "page": {
        "description": "Coaegresco a tandem suffoco. Dens coaegresco sint trado chirographum accedo. Taedium tibi utique.",
        "id": 45,
        "slug": "the-warhol-economy",
        "title": "Book The Warhol Economy",
        "type": "product",
      },
      "pageId": 45,
      "price": "266",
    },
  ],
}
`;
